test/e2e/app_autosync_test.go

```diff
@@ -92,3 +93,61 @@ func TestAutoSyncSelfHealEnabled(t *testing.T) {
 			assert.Empty(t, app.Status.Conditions)
 		})
 }
+
+func TestAutoSyncSelfHealRetryAndRefreshEnabled(t *testing.T) {
+	limits := []int64{
+		100, // Repeat enough times to see we move on to 3rd commit without reaching the limit
+		-1,  // Repeat forever
+	}
+
+	for _, limit := range limits {
+		Given(t).
+			Path(guestbookPath).
+			When().
+			// Correctly configured app should be auto-synced once created
+			CreateFromFile(func(app *Application) {
+				app.Spec.SyncPolicy = &SyncPolicy{
+					Automated: &SyncPolicyAutomated{
+						SelfHeal: true,
```

Comment:
SelfHeal (sync when changes are made in k8s making the app out of sync) is not used in this context since there is a new commit (PatchFile)